vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -474,6 +474,158 @@ def query_metrics_wrapper(self, query_specs):
+                                            hostname=self._hostname,
+                                        )
+                                    else:
+                                        self.log.debug(
+                                            "Skipping metric %s because it is not in the list of metrics to collect",
+                                            given_metric.metricId.label,
+                                        )
+                        if latest_metric_time is None:
+                            latest_metric_time = collect_start_time
+        except Exception as e:
+            # Don't get stuck on a failure to fetch a vsan metric
+            # Ignore them for next pass
+            self.log.warning("Unable to fetch Vsan metrics %s", e)
+
+        if latest_metric_time is not None:
+            self.latest_metric_query = latest_metric_time + dt.timedelta(seconds=1)
+        else:
+            # Let's set `self.latest_metric_query` to `collect_start_time` as safeguard in case no metrics are reported
+            # OR something bad happened (which might happen again indefinitely).
+            self.latest_metric_query = collect_start_time
+
+    def query_vsan_metrics(self, starting_time):
+        # we seek to make 1 call per cluster since the cluster id needs to be passed in the perf manager
+        # {cluster_reference: [cluster_id, host1_id, host2_id, host1disk1_id, host1disk2_id, ...]}
+        cluster_nested_elts = {}
+        entity_ref_ids = {}
+        entity_ref_ids['cluster'] = ['cluster-domclient:', 'vsan-cluster-capacity:']
+        entity_ref_ids['host'] = ['host-domclient:', 'host-cpu:']
+        entity_ref_ids['disk'] = ['capacity-disk:', 'cache-disk:']
+        # {id: {0: type, 1: cluster_name, (optional)2: host_name, (optional)3: disk_id}}
+        id_to_tags = {}
+        for cluster in self.infrastructure_cache.get_mors(vim.ClusterComputeResource):
+            cluster_vsan_config = cluster.configurationEx.vsanConfigInfo
+            # only collect vsan metrics if the cluster has vsan enabled
+            if cluster_vsan_config and cluster_vsan_config.enabled:
+                self.log.debug("Collecting vsan metrics for cluster %s", cluster.name)
+                cluster_uuid = cluster_vsan_config.defaultConfig.uuid
+                cluster_nested_elts[cluster] = [cluster_uuid]
+                id_to_tags[cluster_uuid] = {1: cluster.name, 0: 'cluster'}
+                for host in cluster.host:
+                    host_uuid = host.configManager.vsanSystem.config.clusterInfo.nodeUuid
+                    if cluster_uuid not in cluster_nested_elts.keys():
+                        cluster_nested_elts[cluster] = [cluster_uuid]
+                    cluster_nested_elts[cluster].append(host_uuid)
+                    id_to_tags[host_uuid] = {1: cluster.name, 2: host.name, 0: 'host'}
+                    host_disks = host.configManager.vsanSystem.QueryDisksForVsan()
```

Comment:
can we move the fetching of disks into `api.py`? That way we can enable retry behavior and separate concerns

Reply:
Done.